modal/functions.py

```diff
@@ -292,7 +292,10 @@ async def _map_invocation(
     have_all_inputs = False
     num_inputs = 0
     num_outputs = 0
-    pending_outputs = {}  # map input_id -> next expected gen_index value
+    # Map input_id -> next expected gen_index value, or -1 if complete
+    # Could contain keys which are not yet in the pending_inputs set
+    pending_outputs: Dict[str, int] = {}
+    completed_outputs: Set[str] = set()
```

Comment:
I guess the potential downside is that for huge maps this will keep one entry for every input in memory until the map completes, but in practice that probably won't matter :)

Reply:
Ah, yes it is not ideal .. I can make this a TODO.